src/Controller/MiddlewareController.php

```diff
@@ -0,0 +1,122 @@
+<?php
+/**
+ * Zend Framework (http://framework.zend.com/)
+ *
+ * @link      http://github.com/zendframework/zf2 for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Zend Technologies USA Inc. (http://www.zend.com)
+ * @license   http://framework.zend.com/license/new-bsd New BSD License
+ */
+
+namespace Zend\Mvc\Controller;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\EventManager\EventManagerInterface;
+use Zend\Http\Request;
+use Zend\Mvc\Exception\ReachedFinalHandlerException;
+use Zend\Mvc\Exception\RuntimeException;
+use Zend\Mvc\MvcEvent;
+use Zend\Psr7Bridge\Psr7ServerRequest;
+use Zend\Router\RouteMatch;
+use Zend\Stratigility\Delegate\CallableDelegateDecorator;
+use Zend\Stratigility\MiddlewarePipe;
+
+/**
+ * @internal don't use this in your codebase, or else @ocramius will hunt you down. This is just an internal
+ * @internal hack to make middleware trigger 'dispatch' events attached to the DispatchableInterface identifier.
+ *
+ * Specifically, it will receive a @see MiddlewarePipe, a @see ResponseInterface prototype, and then dispatch
+ * the pipe whilst still behaving like a normal controller. That is needed for any events attached to
+ * the @see \Zend\Stdlib\DispatchableInterface identifier to reach their listeners on any attached
+ * @see \Zend\EventManager\SharedEventManagerInterface
+ */
+final class MiddlewareController extends AbstractController
+{
+    /**
+     * @var MiddlewarePipe
+     */
+    private $pipe;
+
+    /**
+     * @var ResponseInterface
+     */
+    private $responsePrototype;
+
+    public function __construct(
+        MiddlewarePipe $pipe,
+        ResponseInterface $responsePrototype,
```

Comment:
Not particularly proud of this, but PSR-7's immutability makes it quite acceptable to pass in a request as a prototype. Still, this class is designed as "fire once" and "internal" specifically due to this weirdness.